src/deepsparse/pipeline.py

```
@@ -747,6 +804,39 @@ def route_input_to_bucket(
         pass
 
 
+def _initialize_executor_and_workers(
+    batch_size: Optional[int],
+    workers_or_executor: Optional[Union[int, bool, ThreadPoolExecutor]],
+) -> Tuple[Optional[ThreadPoolExecutor], int]:
+    if isinstance(workers_or_executor, ThreadPoolExecutor):
+        num_async_workers = workers_or_executor._max_workers  # noqa
+        executor = workers_or_executor
+    elif isinstance(workers_or_executor, int):
+        num_async_workers = max(1, workers_or_executor)
+        executor = ThreadPoolExecutor(max_workers=num_async_workers)
+    elif isinstance(workers_or_executor, bool):
+        executor = ThreadPoolExecutor() if workers_or_executor else None
+        num_async_workers = executor._max_workers if workers_or_executor else 1  # noqa
+    elif workers_or_executor is None:
+        executor = None if batch_size else ThreadPoolExecutor()
+        num_async_workers = executor._max_workers if batch_size is None else 1  # noqa
+    else:
+        warnings.warn(
+            "Expected an int, bool or ThreadPoolExecutor to run in async mode"
+            f" but got {workers_or_executor} of type {type(workers_or_executor)}"
+        )
```

we should fail if we get an invalid executor. this is a user error